core/src/main/java/org/apache/calcite/rel/core/Project.java

```diff
@@ -362,7 +362,8 @@ protected boolean deepEquals0(@Nullable Object obj) {
 
   @API(since = "1.24", status = API.Status.INTERNAL)
   protected int deepHashCode0() {
-    return Objects.hash(traitSet, input.deepHashCode(), exps, hints);
+    return Objects.hash(traitSet, input.deepHashCode(),
```

Comment:
does this match what happens for other operators?

Reply:
> does this match what happens for other operators?

Maybe other operators need to check deepHashCode. But project's deepHashCode can due to wrong choose. Changing this can find 2 rules bug, but I already fixed them.

Reply:
There are two reasons to check what is done in other similar places:
- if this is a bug, we should try to find out whether the bug appears in other places so we can fix it. Not in this PR, but at least we can file a new issue about it.
- if no other place checks field names, perhaps the intent of the digest is to ignore field names, and then maybe the bug is elsewhere

Reply:
Note that I personally don't know what the intent of the digest is.

Reply:
> Note that I personally don't know what the intent of the digest is.

please see this issue: https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-6852